internal/build/imgsrc/tokens.go

```diff
@@ -0,0 +1,211 @@
+package imgsrc
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/superfly/fly-go"
+	"github.com/superfly/fly-go/tokens"
+	"github.com/superfly/flyctl/gql"
+	"github.com/superfly/flyctl/internal/config"
+	"github.com/superfly/macaroon"
+	"github.com/superfly/macaroon/flyio"
+	"github.com/superfly/macaroon/resset"
+)
+
+const (
+	buildTokenName   = "Temporary flyctl build token"
+	buildTokenExpiry = 1 * time.Hour
+)
+
+func getBuildToken(ctx context.Context, app *fly.AppCompact) (string, error) {
+	tokens := config.Tokens(ctx)
+
+	orgID, err := strconv.ParseUint(app.Organization.InternalNumericID, 10, 64)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse organization ID: %w", err)
+	}
+
+	var token string
+	if len(tokens.UserTokens) > 0 {
+		token, err = getBuildTokenFromUser(ctx, orgID, app.Organization)
+	} else {
+		token, err = getBuildTokenFromMacaroons(orgID, tokens.Macaroons())
+	}
```

Comment:
Yeah, I'm favouring user tokens to copy the existing behaviour from registry, and because the temporary deploy token has fewer moving parts.

That being said, this should be fairly resilient to token expiry? We should be calling `getBuildToken` right before we use the token, and I think `tokens.Macaroons()` will return tokens that won't expire for at least a minute (from what I understand of the pruning/fetching logic). There's maybe some contrived edge case if you disable BuildKit _and_ have a long multi-stage build _and_ use `FROM registry.fly.io`.

Reply:
I don't have the whole flow in my head. Flyctl call rchab and then rchab uses the tokens to talk to registry, right? I was worried about tokens expiring before rchab has a chance to use them.

Reply:
rchab is basically just a proxy to `dockerd`. So "build and push" is flyctl asking rchab to do a build, then flyctl asking rchab to push the image it just created. In a BuildKit world, we don't send tokens with that build request, BuildKit makes a request back to flyctl when it needs them. So rchab should be using the tokens ~immediately after getting them.

The edge case is a non-BuildKit, legacy multi-stage build where flyctl sends tokens at the start of the build, but Docker doesn't pull the images for later stages until much later, because I think legacy builds are fully sequential.

Reply:
Makes sense. Thanks for the context. There's a lot more going on than I realized.